snapcraft/plugins/ruby.py

```diff
@@ -0,0 +1,144 @@
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+"""The ruby plugin is useful for ruby based parts.
+The plugin uses gem to install dependencies from a `Gemfile`.
+This plugin uses the common plugin keywords as well as those for "sources".
+For more information check the 'plugins' topic for the former and the
+'sources' topic for the latter.
+Additionally, this plugin uses the following plugin-specific keywords:
+    - gems:
+      (list)
+      A list of gems to install.
+    - ruby-version:
+      (string)
+      The version of ruby you want this snap to run.
+"""
+import re
+
+import platform
+
+from os import makedirs, environ
+from os.path import exists, join
+
+from snapcraft import BasePlugin
+from snapcraft.sources import Tar
+
+
+class RubyPlugin(BasePlugin):
+
+    @classmethod
+    def schema(cls):
+        schema = super().schema()
+
+        schema['properties']['ruby-version'] = {
+            'type': 'string',
+            'default': '2.3.1'
+        }
+        schema['properties']['gems'] = {
+            'type': 'array',
+            'minitems': 1,
+            'uniqueItems': True,
+            'items': {
+                'type': 'string'
+            },
+            'default': []
+        }
+        return schema
+
+    @classmethod
+    def get_pull_properties(cls):
+        # Inform Snapcraft of the properties associated with pulling. If these
+        # change in the YAML Snapcraft will consider the build step dirty.
+        return ['ruby-version']
+
+    @classmethod
+    def get_build_properties(cls):
+        # Inform Snapcraft of the properties associated with building. If these
+        # change in the YAML Snapcraft will consider the build step dirty.
+        return ['gems']
+
+    def __init__(self, name, options, project):
+        super().__init__(name, options, project)
+
+        self._ruby_version = self.options.ruby_version
+        self._ruby_part_dir = join(self.partdir, 'ruby')
+        self._ruby_download_url = \
+            'https://cache.ruby-lang.org/pub/ruby/ruby-{}.tar.gz'.format(
+                self._ruby_version)
+        self._ruby_tar = Tar(self._ruby_download_url, self._ruby_part_dir)
+        self._gems = self.options.gems
+        self._install_bundler = False
+
+        self.build_packages.extend(['gcc', 'g++', 'make', 'zlib1g-dev',
+                                    'libssl-dev', 'libreadline-dev'])
+
+        version_map = [('2.%d.[0-9]' % i, '2.%d.0' % i) for i in range(5)]
+
+        for version_regex, version_dir in version_map:
+            if re.compile(version_regex).match(self._ruby_version):
+                self._ruby_version_dir = version_dir
+                break
```

Comment:
@kyrofa yeah .... "It's probably better to handle this case on the path discovery side, using globbing (since there should only ever be a single option)" - could you elaborate on what this would look like possibly? Are there other plugins that use this method?

"Otherwise this plugin needs to change if ruby changes versions in any significant way" - you mean if ruby default install procedure decides to not use the minor release version for the lib dir any longer? 

@kyrofa  @kalikiana I felt creating a map was the most simple and maintainable way forward here ... can you suggest how this might be done differently?